python_modules/libraries/dagster-k8s/dagster_k8s/container_context.py

```diff
@@ -0,0 +1,165 @@
+from typing import TYPE_CHECKING, Any, Dict, List, NamedTuple, Optional
+
+from dagster import check
+from dagster.config.validate import process_config
+from dagster.core.errors import DagsterInvalidConfigError
+from dagster.core.storage.pipeline_run import PipelineRun
+from dagster.utils import merge_dicts
+
+if TYPE_CHECKING:
+    from . import K8sRunLauncher
+
+from .job import DagsterK8sJobConfig
+
+
+class K8sContainerContext(
+    NamedTuple(
+        "_K8sContainerContext",
+        [
+            ("image_pull_policy", Optional[str]),
+            ("image_pull_secrets", List[Dict[str, str]]),
+            ("service_account_name", Optional[str]),
+            ("env_config_maps", List[str]),
+            ("env_secrets", List[str]),
+            ("env_vars", List[str]),
+            ("volume_mounts", List[Dict[str, Any]]),
+            ("volumes", List[Dict[str, Any]]),
+            ("labels", Dict[str, str]),
+            ("namespace", Optional[str]),
+        ],
+    )
+):
+    """Encapsulates configuration that can be applied to a K8s job running Dagster code.
+    Can be persisted on a PipelineRun at run submission time based on metadata from the
+    code location and then included in the job's configuration at run launch time or step
+    launch time."""
+
+    def __new__(
+        cls,
+        image_pull_policy: Optional[str] = None,
+        image_pull_secrets: Optional[List[Dict[str, str]]] = None,
+        service_account_name: Optional[str] = None,
+        env_config_maps: Optional[List[str]] = None,
+        env_secrets: Optional[List[str]] = None,
+        env_vars: Optional[List[str]] = None,
+        volume_mounts: Optional[List[Dict[str, Any]]] = None,
+        volumes: Optional[List[Dict[str, Any]]] = None,
+        labels: Optional[Dict[str, str]] = None,
+        namespace: Optional[str] = None,
+    ):
+        return super(K8sContainerContext, cls).__new__(
+            cls,
+            image_pull_policy=check.opt_str_param(image_pull_policy, "image_pull_policy"),
+            image_pull_secrets=check.opt_list_param(image_pull_secrets, "image_pull_secrets"),
+            service_account_name=check.opt_str_param(service_account_name, "service_account_name"),
+            env_config_maps=check.opt_list_param(env_config_maps, "env_config_maps"),
+            env_secrets=check.opt_list_param(env_secrets, "env_secrets"),
+            env_vars=check.opt_list_param(env_vars, "env_vars"),
+            volume_mounts=check.opt_list_param(volume_mounts, "volume_mounts"),
+            volumes=check.opt_list_param(volumes, "volumes"),
+            labels=check.opt_dict_param(labels, "labels"),
+            namespace=check.opt_str_param(namespace, "namespace"),
+        )
+
+    def merge(self, other: "K8sContainerContext") -> "K8sContainerContext":
```

Comment:
Maybe include a small comment about how we try to combine attributes if possible, otherwise the method argument's attribute takes precendence